File: src/main/java/io/github/bot/terminal/application/domains/bot_commands/repository/BotCommandRepository.java
package io.github.bot.terminal.application.domains.bot_commands.repository;


import lombok.RequiredArgsConstructor;
import org.dizitart.no2.objects.ObjectRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

import static org.dizitart.no2.objects.filters.ObjectFilters.eq;

@Repository
@RequiredArgsConstructor
public class BotCommandRepository {

    private final ObjectRepository<BotCommandDetails> db;
    private final ConcurrentMap<String, ReentrantLock> locks = new ConcurrentHashMap<>();

    public void add(BotCommandDetails details) {
        details.setId(UUID.randomUUID().toString());
        db.insert(details);
    }

    public void update(BotCommandDetails details) {
        db.update(details);
    }

    public Optional<BotCommandDetails> findById(String id) {
        return Optional.ofNullable(db.find(eq("id", id)).firstOrDefault());
    }

    public List<BotCommandDetails> findAll() {
        return db.find().toList();
    }

    public void deleteById(String id) {
        transaction(id, (repository) -> {
            if (repository.findById(id).isPresent()) {
                db.remove(eq("id", id));
            }
        });
    }

    public void transaction(String id, Consumer<BotCommandRepository> action) {
        Lock lock = getLock(id);
        lock.lock();
        try {
            action.accept(this);
        } finally {
            lock.unlock();
        }
    }

    private Lock getLock(String id) {
        return locks.computeIfAbsent(id, (k) -> new ReentrantLock());
    }
}

File: src/main/kotlin/io/github/support/terminal/application/domains/bot_commands/repository/BotCommandRepository.kt
package io.github.support.terminal.application.domains.bot_commands.repository

import org.dizitart.no2.objects.ObjectRepository
import org.dizitart.no2.objects.filters.ObjectFilters
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

@Service
class BotCommandRepository(
        val db: ObjectRepository<BotCommandDetails>
) {
    private val locks: ConcurrentMap<String, ReentrantLock> = ConcurrentHashMap()

    fun add(details: BotCommandDetails) {
        db.insert(details)
    }

    fun update(details: BotCommandDetails) {
        db.update(details)
    }

    fun findById(id: String): BotCommandDetails? {
        return db.find(ObjectFilters.eq("id", id)).firstOrNull()
    }

    fun findAll(): List<BotCommandDetails> {
        return db.find().toList()
    }

    fun deleteById(id: String) {
        transaction(id) { repository: BotCommandRepository ->
            repository.findById(id)?.let {
                db.remove(ObjectFilters.eq("id", id))
            }
        }
    }

    fun transaction(id: String, action: (BotCommandRepository) -> Unit) {
        val lock = getLock(id)
        lock.lock()
        try {
            action(this)
        } finally {
            lock.unlock()
        }
    }

    private fun getLock(id: String): Lock {
        return locks.computeIfAbsent(id) { ReentrantLock() }
    }
}